chain/client/src/sync.rs

```diff
@@ -515,47 +519,64 @@ impl BlockSync {
+            }
+        }
+        if requests.is_empty() {
+            return Ok(false);
+        }
 
         let head = chain.head()?;
         let header_head = chain.header_head()?;
 
         let gc_stop_height = chain.runtime_adapter.get_gc_stop_height(&header_head.last_block_hash);
 
-        let request_from_archival = self.archive && request.height < gc_stop_height;
-        let peer = if request_from_archival {
-            let archival_peer_iter = highest_height_peers.iter().filter(|p| p.chain_info.archival);
-            archival_peer_iter.choose(&mut rand::thread_rng())
-        } else {
-            let peer_iter = highest_height_peers.iter();
-            peer_iter.choose(&mut rand::thread_rng())
-        };
+        for request in requests.iter() {
+            let request_from_archival = self.archive && request.height < gc_stop_height;
+            let peer = if request_from_archival {
+                let archival_peer_iter =
+                    highest_height_peers.iter().filter(|p| p.chain_info.archival);
+                archival_peer_iter.choose(&mut rand::thread_rng())
+            } else {
+                let peer_iter = highest_height_peers.iter();
+                peer_iter.choose(&mut rand::thread_rng())
+            };
 
-        if let Some(peer) = peer {
-            debug!(target: "sync", "Block sync: {}/{} requesting block {} from {} (out of {} peers)",
-		   head.height, header_head.height, next_hash, peer.peer_info.id, highest_height_peers.len());
-            self.network_adapter.do_send(PeerManagerMessageRequest::NetworkRequests(
-                NetworkRequests::BlockRequest {
-                    hash: request.hash,
-                    peer_id: peer.peer_info.id.clone(),
-                },
-            ));
-        } else {
-            warn!(target: "sync", "Block sync: {}/{} No available {}peers to request block {} from",
-		  head.height, header_head.height, if request_from_archival { "archival " } else { "" }, next_hash);
+            if let Some(peer) = peer {
+                debug!(target: "sync", "Block sync: {}/{} requesting block {} from {} (out of {} peers)",
+                       head.height, header_head.height, request.hash, peer.peer_info.id, highest_height_peers.len());
+                self.network_adapter.do_send(PeerManagerMessageRequest::NetworkRequests(
+                    NetworkRequests::BlockRequest {
+                        hash: request.hash,
+                        peer_id: peer.peer_info.id.clone(),
+                    },
+                ));
+            } else {
+                warn!(target: "sync", "Block sync: {}/{} No available {}peers to request block {} from",
+                      head.height, header_head.height, if request_from_archival { "archival " } else { "" }, request.hash);
+            }
         }
 
-        self.last_request = Some(request);
+        // we know that requests is non-empty
+        self.last_request = Some(requests.pop().unwrap());
```

Comment:
What should we do if any of the request hit that 'warn' part ? 
(as in such case, the request was never sent, right?)

should we set the last_request to the last one that was actually successfully sent over the network?

Reply:
Good point, changed it